waku/v2/protocol/waku_rln_relay/group_manager/on_chain/group_manager.nim

```diff
@@ -175,6 +179,22 @@ proc parseEvent*(event: type MemberRegistered,
 
 type BlockTable* = OrderedTable[BlockNumber, seq[Membership]]
 
+proc backfillRootQueue*(g: OnchainGroupManager, blockTable: BlockTable): Future[void] {.async.} =
+  if blocktable.len() > 0:
+      for blockNumber, members in blocktable.pairs():
+        let deletionSuccess = g.rlnInstance.removeMembers(members.mapIt(it.index))
+        debug "deleting members to reconcile state"
+        if not deletionSuccess:
+          error "failed to delete members from the tree", success=deletionSuccess
+          raise newException(ValueError, "failed to delete member from the tree, tree is inconsistent")
+      # backfill the tree's acceptable roots
+      for i in 0..blocktable.len()-1:
+        # remove the last root
+        g.validRoots.popLast()
+      for i in 0..blockTable.len()-1:
+        # add the backfilled root
+        g.validRoots.addLast(g.validRootBuffer.popLast())
+
 proc getEvents*(g: OnchainGroupManager, fromBlock: BlockNumber, toBlock: Option[BlockNumber] = none(BlockNumber)): Future[BlockTable] {.async.} =
```

Comment:
The processing under `getEvents` is getting quite complex to follow. Any possibility to extract some of these functionalities to separate steps? I know there's a lot of interaction here between different parts of the logic so it may not be trivial to do.